### ISSUE TITLE
Memory requirement for hyperelastic materials

### ISSUE THREAD

**Issue by meyer-nils** [meyer-nils/torch-fem#29]
Automatic differentiation in hyperelastic materials leads to huge memory consumption, because the computational graph is kept throughout Newton iterations. This is the result of 
```
with profile(
    activities=[ProfilerActivity.CPU],
    profile_memory=True,
    record_shapes=True,
    with_stack=True,
) as prof:
    u3, f3, σ3, F3, α3 = box.solve(
        increments=increments[:2],
        return_intermediate=True,
        nlgeom=True,
        verbose=True,
        rtol=0.5,
    )
prof.export_memory_timeline("memory.html")
```
in the 'solid/large_stretch' example with Neo-Hookean material for only two iterations: 

<img width="1218" height="746" alt="Image" src="https://github.com/user-attachments/assets/3aa37abb-1035-461e-a81b-43689d9435b3" />

The 16 steps come from 2 iterations x 8 integration points.

Things to test: 

- Detach and free graph after each step. However, this will sacrifice differentiability. 
- ...
- ...

**Comment by cptyannick** [comment]
Hi! I had similar issues for optical simulations. I solved this using checkpoints (see [pytorch docs](https://docs.pytorch.org/docs/stable/checkpoint.html)). Might be a nice idea to add an optional argument to solve that enables checkpointing after `n` steps. This would trade memory usage for computation time (in backwards pass). 

Would this approach be acceptable? I think detaching and freeing would be too big a sacrifice, since differentiability is 1) a huge advantage of this project and 2) any user can `with torch.no_grad()` for not storing a computation graph.

**Comment by meyer-nils** [comment]
Hi Yannick, thanks a lot for this suggestion! I will definitely look into this as soon as possible, because I fully agree that sacrificing differentiability defeats the purpose of the whole project.

This issue was originally just a note to myself - seeing your suggestion really made my day. 👍

**Comment by cptyannick** [comment]
Hi Nils, nice to hear! Let me know if you're interested in accepting a pull request for this feature. 😃

**Comment by meyer-nils** [comment]
Sure, absolutely.

**Comment by meyer-nils** [comment]
@cptyannick: Just a short heads-up for you: I switched to a Total Lagrangian Formulation, as this simplifies the hyperelastic step definition and does not need explicit computation of the geometric stiffness. Also, this fixed a couple convergence issues with very large compressions...

For now, it detaches the graph after computing the first Piola Kirchhoff stress and the elasticity tensor: 

```python
# Compute deformation gradient
F_new = F + H_inc
F_new.requires_grad_(True)
# Compute first Piola-Kirchhoff stress tensor
P_new = vmap(jacrev(self.psi))(F_new, self.params)
# Update internal state
state_new = state
# Compute algorithmic tangent stiffness
ddsdde = vmap(jacrev(jacrev(self.psi)))(F_new, self.params)
return P_new.detach(), state_new.detach(), ddsdde.detach()
```

Obviously, this is a big drawback, as the gradient cannot flow through hyperelastic material models. Making this differentiable without enormous memory usage is still an issue.

**Comment by JuliusGulde** [comment]
Hey Nils! I’ve been looking into some modal analysis for my own work and ran into the same "memory vs. differentiability" wall you mentioned here.
I stumbled across a library that seems like a perfect fix for this: torch-sla (from the paper [arXiv:2601.13994](https://arxiv.org/abs/2601.13994)).
It basically acts as an adjoint wrapper for iterative solvers. Instead of unrolling the Newton iterations (which is why your memory is spiking), it uses the Implicit Function Theorem to get the gradients.
Might be worth a look to see if you can swap the current Newton implementation for their Newton solver. It seems to solve the exact trade-off you were wrestling with!

**Comment by meyer-nils** [comment]
Thanks, this looks like a perfect fit. Currently, the `sparse_solve` wrapper in torch-fem does exactly what Algorithm 1 in the paper does, but wrapping the entire Newton-Raphson loop with adjoints is definitely the way to go. I'll certainly give it a try!

**Comment by meyer-nils** [comment]
@cptyannick @JuliusGulde Thanks for the suggestions. I implemented a adjoint newton solve in v0.6.0 and it solve this issue for now. 🎉
